numba/npyufunc/_internal.c

```diff
@@ -104,6 +105,557 @@ PyTypeObject PyUFuncCleaner_Type = {
+    if (!arg_types_array) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+    for (idx = 0; idx < nargs; idx++) {
+        arg_types_array[idx] = (int)PyLong_AsLong(PyList_GET_ITEM(type_list,
+                                                                  idx));
+    }
+    if (PyErr_Occurred()) {
+        PyArray_free(arg_types_array);
+        arg_types_array = NULL;
+    }
+    return arg_types_array;
+}
+
+static PyObject *
+dufunc__add_loop(PyDUFuncObject * self, PyObject * args)
+{
+    PyUFuncObject * ufunc=self->ufunc;
+    unsigned long loop_ptr=0, data_ptr=0;
+    int idx=-1, usertype=NPY_VOID;
+    int *arg_types_arr=NULL;
+    PyObject *arg_types=NULL;
+    PyUFuncGenericFunction old_func=NULL;
+
+    if (self->frozen) {
+        PyErr_SetString(PyExc_ValueError,
+                        "_DUFunc._add_loop() called for frozen dufunc");
+        return NULL;
+    }
+
+    if (!PyArg_ParseTuple(args, "kO!|k", &loop_ptr, &PyList_Type, &arg_types,
+                          &data_ptr)) {
+        return NULL;
+    }
+
+    arg_types_arr = _build_arg_types_array(arg_types, (Py_ssize_t)ufunc->nargs);
+    if (!arg_types_arr) goto _dufunc__add_loop_fail;
+    for (idx = 0; idx < ufunc->nargs; idx++) {
+        if (arg_types_arr[idx] >= NPY_USERDEF) {
+            usertype = arg_types_arr[idx];
+        }
+    }
+
+    if (usertype != NPY_VOID) {
+        if (PyUFunc_RegisterLoopForType(ufunc, usertype,
+                                        (PyUFuncGenericFunction)loop_ptr,
+                                        arg_types_arr, (void *)data_ptr) < 0) {
+            goto _dufunc__add_loop_fail;
+        }
+    } else if (PyUFunc_ReplaceLoopBySignature(ufunc,
+                                              (PyUFuncGenericFunction)loop_ptr,
+                                              arg_types_arr, &old_func) == 0) {
+        /* TODO: Consider freeing any memory held by the old loop (somehow) */
+        for (idx = 0; idx < ufunc->ntypes; idx++) {
+            if (ufunc->functions[idx] == (PyUFuncGenericFunction)loop_ptr) {
+                ufunc->data[idx] = (void *)data_ptr;
+                break;
+            }
+        }
+    } else {
+        int ntypes=ufunc->ntypes + 1;
+        PyUFuncGenericFunction *functions=NULL;
+        void **data=NULL;
+        char *types=NULL;
+        void *newptr=NULL, *oldptr=NULL;
+        size_t functions_size=sizeof(PyUFuncGenericFunction) * ntypes;
+        size_t data_size=sizeof(void *) * ntypes;
+        size_t type_ofs=sizeof(char) * ufunc->ntypes * ufunc->nargs;
+        size_t newsize=(functions_size + data_size +
+                        (sizeof(char) * ntypes * ufunc->nargs) +
+                        (sizeof(void *) - 1)); /* Alignment fudge... */
+
+        oldptr = ufunc->ptr;
+        newptr = PyArray_malloc(newsize);
+        functions =
+          (PyUFuncGenericFunction*)((char *)newptr +
+                                    ((size_t)newptr % sizeof(void *)));
```

Comment:
By the way here is what POSIX says about malloc:
http://pubs.opengroup.org/onlinepubs/9699919799/functions/malloc.html
> The pointer returned if the allocation succeeds shall be suitably aligned so that it may be assigned to a pointer to any type of object and then used to access such an object in the space allocated

Here is what MSDN says:
https://msdn.microsoft.com/en-us/library/aa246461%28v=vs.60%29.aspx?f=255&MSPPError=-2147217396
> The storage space pointed to by the return value is guaranteed to be suitably aligned for storage of any type of object.